### PR TITLE
Add Teleport multi-region blueprint

### PR DESCRIPTION
**Pull request by hugoShaka**
This PR adds documentation on how to create a multi-region Teleport cluster.

Depends on https://github.com/gravitational/teleport/pull/42023

### CODE REVIEW

File: docs/pages/deploy-a-cluster/multi-region-blueprint.mdx [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+Updating Teleport on multi-region setups poses several challenges.
+Strictly following Teleport compatibility guarantees requires downscaling to a
+single auth before updating. In a multi region setup, this requires:
+
+- relying on a single region for auth during the udpate
+- suspending two regions, ot allowing proxies to connect to auths from a
+  different region (this is not the case by default and requires a Cilium
+  cross-cluster service and a custom join token for the Proxy Service).
+
+<Admonition type="note">
```
Comment (klizhentas) [inline docs/pages/deploy-a-cluster/multi-region-blueprint.mdx:213]: I'd be more prescriptive here and say:

You can perform rolling upgrades, when there are multiple auth servers of different versions running at the same time.

For example, let's assume there are multiple versions of Teleport available:

```
14.5.3
14.5.4
14.5.6
14.6.1
14.6.2
15.0.0
15.0.1
```

* You upgrade between **any** patch versions of the same minor and major release, e.g. `14.5.3` to `14.5.6`
* You upgrade between successive minor versions, e.g. `14.5.3` to `14.6.1`
* You upgrade between strictly previous minor version and next major version, for example `14.6.1` to `15.0.1`.

In all other cases, the results of a rolling upgrade may result in data corruption and outage.

**Review by klizhentas — COMMENTED** [review]
Overall, beautiful doc, excited to see it published thanks @hugoShaka .

I think I'd add notes on how to configure multi-region S3 storage and what scaling/replication settings to use on cockroach db if any.

File: docs/pages/deploy-a-cluster/multi-region-blueprint.mdx (hunk 1 above)
Comment (hugoShaka) [reply docs/pages/deploy-a-cluster/multi-region-blueprint.mdx:213]: I tried to come up with such rules but according to @rosstimothy we can lose data even by rolling update between two patches. [Slack link](https://gravitational.slack.com/archives/C03PJTF5BPD/p1717014431273299?thread_ts=1715189733.223619&cid=C03PJTF5BPD). We're just accepting the risk in cloud as we're following the version very closely and the probability is low, but the risk is always non-null.

File: docs/pages/deploy-a-cluster/multi-region-blueprint.mdx [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+```cockroach
+ALTER DATABASE teleport_backend SURVIVE ZONE FAILURE;
+ALTER DATABASE teleport_audit SURVIVE ZONE FAILURE;
+```
+
+<Admonition type="note">
+    CockraochDB core does not offer the regional features CockraochDB enterprise
+    does. The performance might be impacted. CockroachDB core is not recommended
+    for production in multi-region setups.
+</TabItem>
```
Comment (hugoShaka) [inline docs/pages/deploy-a-cluster/multi-region-blueprint.mdx:149]:
```suggestion
</Admonition>
</TabItem>
```

File: docs/pages/deploy-a-cluster/multi-region-blueprint.mdx [hunk 3]
```diff
@@ -0,0 +1,314 @@
+---
+title: "Multi-region blueprint"
```
Comment (webvictim) [inline docs/pages/deploy-a-cluster/multi-region-blueprint.mdx:2]:
```suggestion
title: "Multi-region Blueprint"
```

File: docs/pages/deploy-a-cluster/introduction.mdx [hunk 4]
```diff
@@ -35,3 +35,8 @@ Teleport cluster on your cloud:
 - [AWS Multi-Region Proxy Deployment](./deployments/aws-gslb-proxy-peering-ha-deployment.mdx)
 - [Google Cloud](./deployments/gcp.mdx)
 - [IBM Cloud](./deployments/ibm.mdx)
+
+## Deploying a multi-region Teleport cluster
+
+The [multi-region Teleport blueprint](./multi-region-blueprint.mdx) describes
```
Comment (webvictim) [inline docs/pages/deploy-a-cluster/introduction.mdx:41]:
```suggestion
The [multi-region Teleport Blueprint](./multi-region-blueprint.mdx) describes
```

File: docs/pages/deploy-a-cluster/multi-region-blueprint.mdx [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+  For example, Amazon SLAs are 99.99% for both DynamoDB and EC2.
+- This blueprint relies on many complex components. If you don't have the team
+  capacity and expertise to deploy and operate such components, you will end
+  up with a lower availability than a single-region multi-zone setup.
+- You are responsible for deploying and maintaining the Teleport dependencies
+  including the database, object storage, container runtime, network plugins, ...
+
+<Admonition type="tip">
+
+For the teams with limited bandwidth who need multi-region clusters, Teleport
```
Comment (webvictim) [inline docs/pages/deploy-a-cluster/multi-region-blueprint.mdx:29]:
```suggestion
For teams with limited bandwidth who need multi-region clusters, Teleport
```

File: docs/pages/deploy-a-cluster/multi-region-blueprint.mdx [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+offers to host cross-region clusters. You can benefit from the
+setup described in this document without the operational costs and complexity,
+see [the Teleport Enterprise Cloud page](../choose-an-edition/teleport-cloud/introduction.mdx)
```
[the remark this answers is not in the document]
Comment (webvictim) [reply docs/pages/deploy-a-cluster/multi-region-blueprint.mdx:32]: I think this is being updated to "managed Teleport Enterprise" soon which reads better, but for now we have to make the linter happy.

File: docs/pages/deploy-a-cluster/multi-region-blueprint.mdx [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+for more details.
+
+</Admonition>
+
+## Architecture overview
+
+A multi-region Teleport deployment is composed of several regional Teleport
+deployments. Teleport stores its state in a multi-region backend and uses
+[Proxy Peering](../architecture/proxy-peering.mdx) to connect users to resources
+hosted a different region.
```
Comment (webvictim) [inline docs/pages/deploy-a-cluster/multi-region-blueprint.mdx:42]:
```suggestion
hosted in a different region.
```

File: docs/pages/deploy-a-cluster/multi-region-blueprint.mdx [hunk 8]
```diff
@@ -0,0 +1,314 @@
...
+
+The recommended multi-region backend
```
Comment (webvictim) [inline docs/pages/deploy-a-cluster/multi-region-blueprint.mdx:44]: Isn't this the only supported multi-region backend?

File: docs/pages/deploy-a-cluster/multi-region-blueprint.mdx [hunk 9]
```diff
@@ -0,0 +1,314 @@
...
+
+- **3 regions**. Teleport only needs 2, but CockroachDB needs 3 for its quorum mechanism.
+- Teleport Proxy Service instances able to dial each other by IP address.
+  This means you have **cross-region Pod/Instance connectivity**. This is typically
+  achieved with VPC peering and/or service mesh.
+- A **multi-region object storage** for session recordings.
+- **GeoDNS** to route users and Teleport agents to the closest Teleport cluster.
+- A **multi-region CockroachDB cluster**. You can use the CockroachDB core
+  features, but optimal performance and fault-tolerance is achieved by using
+  [the multi-region features from CockroachDB Enterprise](https://www.cockroachlabs.com/docs/stable/enterprise-licensing).
```
Comment (webvictim) [inline docs/pages/deploy-a-cluster/multi-region-blueprint.mdx:60]: I think that as we say CockroachDB core is not recommended for multi-region production use below, we should be explicit about that here too.

File: docs/pages/deploy-a-cluster/multi-region-blueprint.mdx [hunk 10]
```diff
@@ -0,0 +1,314 @@
...
+
+## Implementing multi-region Teleport on Kubernetes
+
+Teams with Kubernetes infrastructure will want to host Teleport in Kubernetes.
+This section describes this setup with more details.
```
Comment (webvictim) [inline docs/pages/deploy-a-cluster/multi-region-blueprint.mdx:65]:
```suggestion
This section describes this setup in more detail.
```

File: docs/pages/deploy-a-cluster/multi-region-blueprint.mdx [hunk 11]
```diff
@@ -0,0 +1,314 @@
...
+
+### Teleport Dependencies
+
+- Create 3 peered regional VPCs (e.g., on AWS) or 1 global VPC (e.g., on GKE).
+- Create 1 Kubernetes cluster in each region, **Pod CIDRs must not overlap**.
+- Ensure pod mesh connectivity. Proxy pods must be able to dial each other
+  [on port 3021](../reference/networking.mdx#ports). The setup will vary
+  based on your network and CNI plugin configuration:
+    - If pods are using the native network layer (GKE, AWS vpc-cni addon, ...)
```
Comment (webvictim) [inline docs/pages/deploy-a-cluster/multi-region-blueprint.mdx:74]:
```suggestion
    - If pods are using the native network layer (for example GKE, or AWS' vpc-cni addon)
```

File: docs/pages/deploy-a-cluster/multi-region-blueprint.mdx [hunk 12]
```diff
@@ -0,0 +1,314 @@
...
+      connect pods from each Kubernetes cluster.
+    - If pods are given non-native IP addresses, you can use the Cilium
+      cluster-mesh capabilities to connect your pods. They
+      provide [a step-by-step guide on how to mesh clusters](https://docs.cilium.io/en/stable/network/clustermesh/clustermesh/).
+- Set up a multi region object storage:
+    - on GCP, GCS has a multi-region location
+    - on-prem, MinIO supports [multi-site bucket replication](https://min.io/docs/minio/linux/administration/bucket-replication/enable-server-side-multi-site-bucket-replication.html)
+    - on AWS, you can [set up two-way-replication](https://docs.aws.amazon.com/AmazonS3/latest/userguide/mrap-create-two-way-replication-rules.html)
+      between S3 buckets.
+- Setup a CockroachDB multi-region cluster
```
Comment (webvictim) [inline docs/pages/deploy-a-cluster/multi-region-blueprint.mdx:85]: Maybe link to https://www.cockroachlabs.com/docs/v24.1/multiregion-overview ?

File: docs/pages/deploy-a-cluster/multi-region-blueprint.mdx [hunk 13]
```diff
@@ -0,0 +1,314 @@
...
+  ```sql
+  CREATE DATABASE teleport_backend;
+  CREATE DATABASE teleport_audit;
+  CREATE USER teleport;
+  GRANT CREATE ON DATABASE teleport_backend TO teleport;
+  GRANT CREATE ON DATABASE teleport_audit TO teleport;
+  SET CLUSTER SETTING kv.rangefeed.enabled = true;
+  ```
+
+Then sign a certificate for the `telpeort` user
```
Comment (webvictim) [inline docs/pages/deploy-a-cluster/multi-region-blueprint.mdx:99]:
```suggestion
Then sign a certificate for the `teleport` user
```

File: docs/pages/deploy-a-cluster/multi-region-blueprint.mdx [hunk 14]
```diff
@@ -0,0 +1,314 @@
...
+with [the `cockroach certs` command](https://www.cockroachlabs.com/docs/stable/cockroach-cert).
+
+You must end up with 3 files:
+- `client.teleport.crt`
+- `client.teleport.key`
+- `ca.crt`
+
+<Tabs>
+<TabItem label="CockroachDB enterprise">
```
Comment (webvictim) [inline docs/pages/deploy-a-cluster/multi-region-blueprint.mdx:108]:
```suggestion
<TabItem label="CockroachDB Enterprise">
```

File: docs/pages/deploy-a-cluster/multi-region-blueprint.mdx [hunk 15]
```diff
@@ -0,0 +1,314 @@
...
+For CockroachDB enterprise you must declare both zones and regions to
```
Comment (webvictim) [inline docs/pages/deploy-a-cluster/multi-region-blueprint.mdx:109]:
```suggestion
For CockroachDB Enterprise you must declare both zones and regions to
```

File: docs/pages/deploy-a-cluster/multi-region-blueprint.mdx [hunk 16]
```diff
@@ -0,0 +1,314 @@
...
+
+ALTER DATABASE teleport_audit SET PRIMARY REGION <region1>;
+ALTER DATABASE teleport_audit ADD REGION IF NOT EXISTS <region2>;
+ALTER DATABASE teleport_audit SET SECONDARY REGION <region2>;
+ALTER DATABASE teleport_audit SURVIVE REGION FAILURE;
+```
+
+<Admonition type="note" title="Impact of the CockroachDB regions">
+    CockroachDB stores the Teleport data in two regions. If you only deploy two
+    regional Teleport clusters, pick the same regions. IF you deploy 3 Teleport
```
Comment (webvictim) [inline docs/pages/deploy-a-cluster/multi-region-blueprint.mdx:126]:
```suggestion
    regional Teleport clusters, pick the same regions. If you deploy 3 Teleport
```

File: docs/pages/deploy-a-cluster/multi-region-blueprint.mdx [hunk 17]
```diff
@@ -0,0 +1,314 @@
...
+
+```sql
+ALTER DATABASE teleport_backend SURVIVE ZONE FAILURE;
+ALTER DATABASE teleport_audit SURVIVE ZONE FAILURE;
+```
+
+<Admonition type="note">
+    CockroachDB core does not offer the regional features CockroachDB enterprise
+    does. The performance might be impacted. CockroachDB core is not recommended
+    for production in multi-region setups.
```
Comment (webvictim) [inline docs/pages/deploy-a-cluster/multi-region-blueprint.mdx:148]:
```suggestion
    CockroachDB core does not offer the regional features CockroachDB Enterprise
    does. The performance might be impacted. CockroachDB core is not recommended
    for production in multi-region setups.
```

File: docs/pages/deploy-a-cluster/multi-region-blueprint.mdx [hunk 18]
```diff
@@ -0,0 +1,314 @@
...
+</Admonition>
+</TabItem>
+</Tabs>
+
+
+### Deploying Teleport
+
+Once all Teleport dependencies are set up, you can deploy Teleport via the
+`teleport-cluster` Helm chart. You need to create one release per Kubernetes cluster.
+The values for a specific region look like:
```
Comment (webvictim) [inline docs/pages/deploy-a-cluster/multi-region-blueprint.mdx:158]:
```suggestion
Here is an example of the values for one specific region:
```

File: docs/pages/deploy-a-cluster/multi-region-blueprint.mdx [hunk 19]
```diff
@@ -0,0 +1,314 @@
...
+      readOnly: true
+  extraEnv:
+    - name: PGSSLROOTCERT
+      value: /var/lib/db-certs/ca.crt
+    - name: PGSSLCERT
+      value: /var/lib/db-certs/client.teleport.crt
+    - name: PGSSLKEY
+      value: /var/lib/db-certs/client.teleport.key
+
+# Pass the TLS certificate use by the proxy
```
Comment (webvictim) [inline docs/pages/deploy-a-cluster/multi-region-blueprint.mdx:205]:
```suggestion
# Pass the TLS certificate to be used by the proxy
```

File: docs/pages/deploy-a-cluster/multi-region-blueprint.mdx [hunk 20]
```diff
@@ -0,0 +1,314 @@
...
+routed to the closest working Teleport deployment.
+
+## Update strategy
+
+Updating Teleport on multi-region setups poses several challenges.
+Strictly following Teleport compatibility guarantees requires downscaling to a
+single auth before updating. In a multi-region setup, this requires:
+
+- relying on a single region for auth during the update
+- suspending two regions, ot allowing proxies to connect to auths from a
```
Comment (webvictim) [inline docs/pages/deploy-a-cluster/multi-region-blueprint.mdx:231]:
```suggestion
- suspending two regions, or allowing proxies to connect to auths from a
```

File: docs/pages/deploy-a-cluster/multi-region-blueprint.mdx [hunk 21]
```diff
@@ -0,0 +1,314 @@
...
+  different region (this is not the case by default and requires a Cilium
+  cross-cluster service and a custom join token for the Proxy Service).
+
+<Admonition type="note">
+
+    It's possible to do Auth rolling updates. In such case, multiple auth
+    versions might be running concurrently. Newly introduced resource fields
+    might not be properly set until all auth are rolled out.
```
Comment (webvictim) [inline docs/pages/deploy-a-cluster/multi-region-blueprint.mdx:239]:
```suggestion
    might not be properly set until all auths are rolled out.
```

File: docs/pages/deploy-a-cluster/multi-region-blueprint.mdx [hunk 22]
```diff
@@ -0,0 +1,314 @@
...
+    This consistency/usability tradeoff can be acceptable if the rollout period
+    is short and the version difference is minimal.
+
+</Admonition>
+
+## CockroachDB sizing
+
+A 3x3 CockroachDB cluster (1 node per zone, 3 zones per region, 3 regions)
+whose nodes have 2 CPUs and 8GiB of memory can sustain a Teleport cluster with
+10 000 SSH nodes.
```
Comment (webvictim) [inline docs/pages/deploy-a-cluster/multi-region-blueprint.mdx:250]:
```suggestion
10,000 SSH nodes.
```

File: docs/pages/deploy-a-cluster/multi-region-blueprint.mdx [hunk 23]
```diff
@@ -0,0 +1,314 @@
...
+The CockroachDB performance depends on the storage type. Make sure to store
+CockroachDB's state on SSDs. When running on Kubernetes, the default storage
+class might not be the one offering the best latency and throughput.
+
+See the [Teleport scaling](../management/operations/scaling.mdx) page for more
+details on how to size Teleport.
+
+Teleport load varies depending on the usage type (number of connected
+nodes, number of protected resources, number of roles, concurrent connections,
+user activity, ...), so you will need to test and adapt the sizing based on your
```
Comment (webvictim) [inline docs/pages/deploy-a-cluster/multi-region-blueprint.mdx:261]:
```suggestion
and user activity), so you will need to test and adapt the sizing based on your
```

File: docs/pages/deploy-a-cluster/multi-region-blueprint.mdx [hunk 24]
```diff
@@ -0,0 +1,314 @@
...
+specific needs.
+
+
+<Admonition type="important">
+    When running on Kubernetes, unless you are using [the static CPU
+    policy](https://kubernetes.io/docs/tasks/administer-cluster/cpu-management-policies/#static-policy)
+    we don't recommend setting CPU limits to Teleport nor CockroachDB pods.
```
Comment (webvictim) [inline docs/pages/deploy-a-cluster/multi-region-blueprint.mdx:268]:
```suggestion
    we don't recommend setting CPU limits on either Teleport or CockroachDB pods.
```

File: docs/pages/deploy-a-cluster/multi-region-blueprint.mdx [hunk 25]
```diff
@@ -0,0 +1,314 @@
...
+
+    Due to the way CPU throttling is implemented, multi-threaded I/O-intensive
+    applications such as Teleport and CockroachDB will degrade as soon as they
+    reach the CPU limit. This will lead to a service disruption and very likely
+    to a complete outage.
+
+    As rule of thumb, most critical pod `resources` section should look like this:
```
Comment (webvictim) [inline docs/pages/deploy-a-cluster/multi-region-blueprint.mdx:275]:
```suggestion
    As a rule of thumb, most critical pods' `resources` section should look like this:
```

File: docs/pages/deploy-a-cluster/multi-region-blueprint.mdx [hunk 26]
```diff
@@ -0,0 +1,314 @@
...
+Terraform. With 3 buckets `bucketA`, `bucketB`, `bucketC` you need 6 rules:
+- from `bucketA` to `bucketB`
+- from `bucketA` to `bucketC`
+- from `bucketB` to `bucketA`
+- from `bucketB` to `bucketC`
+- from `bucketC` to `bucketA`
+- from `bucketC` to `bucketB`
+
+<Admonition type="note">
+    If you want Teleport to automatically cleanup session recordings, you must
```
Comment (webvictim) [inline docs/pages/deploy-a-cluster/multi-region-blueprint.mdx:312]:
```suggestion
    If you want Teleport to automatically clean up session recordings, you must
```

File: docs/pages/deploy-a-cluster/multi-region-blueprint.mdx [hunk 27]
```diff
@@ -0,0 +1,314 @@
...
+    to enable "Delete marker replication" in the replication rules.
```
Comment (webvictim) [inline docs/pages/deploy-a-cluster/multi-region-blueprint.mdx:313]:
```suggestion
    enable "Delete marker replication" in the replication rules.
```

File: docs/pages/reference/backends.mdx [hunk 28]
```diff
@@ -12,10 +12,10 @@ read/write ratio, mutability, etc.).
 
 | Data type | Description | Supported storage backends |
 | - | - | - |
-| core cluster state | Cluster configuration (e.g. users, roles, auth connectors) and identity (e.g. certificate authorities, registered nodes, trusted clusters). | Local directory (SQLite), etcd, PostgreSQL, AWS DynamoDB, GCP Firestore |
+| core cluster state | Cluster configuration (e.g. users, roles, auth connectors) and identity (e.g. certificate authorities, registered nodes, trusted clusters). | Local directory (SQLite), etcd, PostgreSQL, AWS DynamoDB, GCP Firestore, CockroachDB |
```
Comment (webvictim) [inline docs/pages/reference/backends.mdx:15]:
```suggestion
| core cluster state | Cluster configuration (e.g. users, roles, auth connectors) and identity (e.g. certificate authorities, registered nodes, trusted clusters). | Local directory (SQLite), etcd, PostgreSQL, Amazon DynamoDB, GCP Firestore, CockroachDB |
```

Comment (webvictim) [reply docs/pages/reference/backends.mdx:15]: 🙄

File: docs/pages/deploy-a-cluster/multi-region-blueprint.mdx (hunk 6 above)
[the remark this answers is not in the document]
Comment (hugoShaka) [reply docs/pages/deploy-a-cluster/multi-region-blueprint.mdx:32]: I mean, that's the name of the destination page. I'm not sure this error applies here. Cloud hosting is the main topic, why would it be in parenthesis?

As this linter is not required to pass I'm happy to ignore it.

File: docs/pages/deploy-a-cluster/multi-region-blueprint.mdx (hunk 8 above)
Comment (hugoShaka) [reply docs/pages/deploy-a-cluster/multi-region-blueprint.mdx:44]: It's complex. Firestore is multi-region already, and you can definitely achieve multi-region with etcd (althought the latency 🤢) , and with postgres if you know the DB enough.

**Comment by klizhentas** [comment]
@hugoShaka can we get this merged?

**Review by r0mant — APPROVED** [review]
lgtm with Gus' comments addressed

**Comment by hugoShaka** [comment]
> @hugoShaka can we get this merged?

I need a second approval

**Review by webvictim — APPROVED** [review]

**Pull request merged**